pallets/bonds/src/lib.rs

```diff
@@ -0,0 +1,340 @@
+
+			// not covered in the tests. Create an asset with empty name should always work
+			let bond_asset_id = T::AssetRegistry::create_bond_asset(ed)?;
+
+			let fee = T::ProtocolFee::get().mul_ceil(amount); // TODO: check
+			let amount_without_fee = amount.checked_sub(&fee).ok_or(ArithmeticError::Overflow)?;
+			let pallet_account = Self::pallet_account_id();
+
+			T::Currency::transfer(asset_id, &who, &pallet_account, amount_without_fee)?;
+			T::Currency::transfer(asset_id, &who, &T::FeeReceiver::get(), fee)?;
+			T::Currency::deposit(bond_asset_id, &who, amount_without_fee)?;
+
+			Bonds::<T>::insert(
+				bond_asset_id,
+				Bond {
+					maturity,
+					asset_id,
+					amount: amount_without_fee,
+				},
+			);
+
+			Self::deposit_event(Event::BondTokenCreated {
+				issuer: who,
+				asset_id,
+				bond_asset_id,
+				amount: amount_without_fee,
+				fee,
+			});
+
+			Ok(())
+		}
+
+		/// Redeem bonds for the underlying asset.
+		/// The amount of the underlying asset the `origin` receives is 1:1 to the `amount` of the bonds.
+		/// Anyone who holds the bonds is able to redeem them.
+		/// The bonds can be partially redeemed.
+		///
+		/// Parameters:
+		/// - `origin`: account id
+		/// - `asset_id`: bond asset id
+		/// - `amount`: the amount of the bonds to redeem for the underlying asset
+		///
+		/// Emits `BondsRedeemed` event when successful.
+		///
+		#[pallet::call_index(1)]
+		#[pallet::weight(<T as Config>::WeightInfo::redeem())]
+		pub fn redeem(origin: OriginFor<T>, bond_id: T::AssetId, amount: T::Balance) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+
+			Bonds::<T>::try_mutate_exists(bond_id, |maybe_bond_data| -> DispatchResult {
+				let bond_data = maybe_bond_data.as_mut().ok_or(Error::<T>::BondNotRegistered)?;
+
+				let now = T::TimestampProvider::now();
+				ensure!(now >= bond_data.maturity, Error::<T>::BondNotMature);
+
+				T::Currency::withdraw(bond_id, &who, amount)?;
+
+				bond_data.amount = bond_data.amount.checked_sub(&amount).ok_or(ArithmeticError::Overflow)?;
+
+				let pallet_account = Self::pallet_account_id();
+				T::Currency::transfer(bond_data.asset_id, &pallet_account, &who, amount)?;
+
+				// if there are no bonds left, remove the bonds from the storage
+				if bond_data.amount.is_zero() {
+					*maybe_bond_data = None;
+				}
+
+				Self::deposit_event(Event::BondsRedeemed { who, bond_id, amount });
+
+				Ok(())
+			})?;
+
+			Ok(())
+		}
+
+		/// Unlock bonds by making them mature.
+		/// The maturity of the bonds is not updated if the bonds are already mature.
+		///
+		/// Parameters:
+		/// - `origin`: needs to be `T::UnlockOrigin`
+		/// - `bond_id`: the asset id of the bonds
+		///
+		/// Emits `BondsUnlocked` event when successful.
+		///
+		#[pallet::call_index(2)]
+		#[pallet::weight(<T as Config>::WeightInfo::redeem())]
+		pub fn unlock(origin: OriginFor<T>, bond_id: T::AssetId) -> DispatchResult {
+			T::UnlockOrigin::ensure_origin(origin)?;
```

Comment:
noone can unlock bonds, bonds are unlocked only by maturity, was this feature requested?

Reply:
also the implementation is wrong, as you effectivelly transform one bond to another by changing its maturity

Reply:
No this feature wasn't requested. But I think it might be helpful when someone accidentally locks some significant amount of tokens with wrongly set maturity. But we can remove it if you want.

Reply:
removed